tmt/steps/provision/__init__.py

```diff
@@ -1918,45 +2045,147 @@ def go(self, force: bool = False) -> None:
 
         # Provision guests
         self._guests = []
-        save = True
-        self.is_multihost = sum(isinstance(phase, ProvisionPlugin) for phase in self.phases()) > 1
-        try:
-            for phase in self.phases(classes=(Action, ProvisionPlugin)):
-                try:
-                    if isinstance(phase, Action):
-                        phase.go()
 
-                    elif isinstance(phase, ProvisionPlugin):
-                        phase.go()
+        def _run_provision_phases(
+                phases: list[ProvisionPlugin[ProvisionStepData]]
+                ) -> tuple[list[ProvisionTask], list[ProvisionTask]]:
+            queue: ProvisionQueue = ProvisionQueue(
+                'provision',
+                self._logger.descend(logger_name=f'{self}.queue'))
 
-                        guest = phase.guest()
-                        if guest:
-                            guest.show()
+            queue.enqueue(phases=phases, logger=queue._logger)
 
-                    if self.is_multihost:
-                        self.info('')
-                except (tmt.utils.RunError, tmt.utils.ProvisionError) as error:
-                    self.fail(str(error))
-                    raise
-                finally:
-                    if isinstance(phase, ProvisionPlugin):
-                        guest = phase.guest()
-                        if guest and (guest.is_ready or self.is_dry_run):
-                            self._guests.append(guest)
+            all_tasks: list[ProvisionTask] = []
+            failed_tasks: list[ProvisionTask] = []
 
-            # Give a summary, update status and save
-            self.summary()
-            self.status('done')
-        except (SystemExit, tmt.utils.SpecificationError) as error:
-            # A plugin will only raise SystemExit if the exit is really desired
-            # and no other actions should be done. An example of this is
-            # listing available images. In such case, the workdir is deleted
-            # as it's redundant and save() would throw an error.
-            save = False
-            raise error
-        finally:
-            if save:
-                self.save()
+            for outcome in queue.run():
+                all_tasks.append(outcome)
+
+                if outcome.exc:
+                    outcome.logger.fail(str(outcome.exc))
+
+                    failed_tasks.append(outcome)
+
+                    continue
+
+                guest = outcome.guest
+
+                if guest:
+                    guest.show()
+
+                    if guest.is_ready or self.is_dry_run:
+                        self._guests.append(guest)
+
+            return all_tasks, failed_tasks
+
+        def _run_action_phases(phases: list[Action]) -> tuple[list[ActionTask], list[ActionTask]]:
+            queue: PhaseQueue[ProvisionStepData] = PhaseQueue(
+                'finish',
```

Comment:
Should this be rather `action` or something like this?

Reply:
Renamed in db0510dc, to `provision.action` and `provision.provision`.